base/io.jl

```diff
@@ -475,10 +475,32 @@ read(filename::AbstractString, args...) = open(io->read(io, args...), convert(St
 read(filename::AbstractString, ::Type{T}) where {T} = open(io->read(io, T), convert(String, filename)::String)
 
 """
-    read!(stream::IO, array::AbstractArray)
-    read!(filename::AbstractString, array::AbstractArray)
+    read!(stream::IO, a::AbstractArray)
+    read!(stream::IO, a::Ref)
+    read!(filename::AbstractString, a::Union{AbstractArray, Ref})
+
+Read `sizeof(a)` bytes of binary data from an I/O stream or file,
```

Comment:
Does the function even work for non-isbits?

Reply:
Yes, we will probably have to distinguish in the docstring both cases of `isbitstype(T)`. (This should probably also be done for a docstring covering `write(a::AbstractArray{T,N})`)

If `isbitstype(T)==false`, then `read!(..., a:AbstractArray{T,N})` just runs
```
for i in eachindex(a)
    a[i] = read(s, T)
end
```
i.e. falls back to whatever `read()` method was provided for type `T`, and the number of bytes read will depend on what that does. Also, it seems `read!(..., a:Ref{T})` currently doesn't have such a fallback and therefore is only intended for `isbitstype(T)==true`.

Reply:
Right, that's what I mean - non-bitstypes can't really be read like that, because of the pointers contained within. They need to be properly deserialized and recreated in a more complicated manner.